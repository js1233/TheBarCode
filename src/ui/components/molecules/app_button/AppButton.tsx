import { FONT_SIZE, SPACE } from "config";
import Colors from "config/Colors";
import { usePreferredTheme } from "hooks";
import React, { useEffect, useState } from "react";
import {
  ActivityIndicator,
  ImageStyle,
  StyleProp,
  StyleSheet,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle
} from "react-native";
import {
  AppLabel,
  TEXT_TYPE
} from "ui/components/atoms/app_label/AppLabel";
import { SvgProp } from "utils/Util";

export interface AppButtonProps extends TouchableOpacityProps {
  onPress?: () => void;
  text?: string;
  buttonStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  textDisableStyle?: StyleProp<TextStyle>;
  textContainerStyle?: StyleProp<ViewStyle>;
  shouldShowProgressBar?: boolean;
  loaderSize?: number;
  loaderColor?: string;
  leftIcon?: SvgProp | undefined;
  rightIcon?: SvgProp;
  buttonType?: BUTTON_TYPES;
  isDisable?: boolean;
  buttonDisableStyle?: StyleProp<ViewStyle>;
  iconStyle?: StyleProp<ImageStyle>;
  shouldShowError?: boolean;
  textType?: TEXT_TYPE;
  shouldNotOptimize?: boolean;
  shouldAlignTextWithLeftIconWithFullWidth?: boolean;
}

export enum BUTTON_TYPES {
  NORMAL = "normal",
  BORDER = "border",
  DASH = "dashed"
}

export const AppButton = React.memo<AppButtonProps>(
  ({
    text,
    onPress,
    iconStyle,
    buttonStyle,
    textStyle,
    textDisableStyle,
    shouldShowProgressBar = false,
    loaderSize = 15,
    loaderColor,
    leftIcon,
    rightIcon,
    buttonType = BUTTON_TYPES.NORMAL,
    isDisable = false,
    buttonDisableStyle,
    shouldShowError = false,
    textType = TEXT_TYPE.NORMAL,
    shouldAlignTextWithLeftIconWithFullWidth = false,
    textContainerStyle
  }) => {
    const theme = usePreferredTheme();
    const [customDisabled, setIsDisabled] = useState<boolean>(isDisable);

    useEffect(() => {
      let timeout: any;
      if (isDisable) {
        setIsDisabled(isDisable);
      } else {
        //when we need to close screen after button click, changing isDisable to false
        //trigger a second click due to whic hapi is called again to prevent this {isDisbaled:false}
        //is calling after a little delay, to properly close the parent screen in the meantime
        timeout = setTimeout(() => {
          setIsDisabled(isDisable);
        }, 300);
      }

      () => clearTimeout(timeout);
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isDisable]);

    const getButtonStyle = () => {
      if (buttonType === BUTTON_TYPES.DASH) {
        return [
          style.buttonWithDash,
          { borderColor: theme.themedColors.primaryColor }
        ];
      } else if (buttonType === BUTTON_TYPES.BORDER) {
        return [
          style.buttonWithBorder,
          {
            borderColor: shouldShowError
              ? Colors.colors.red
              : theme.themedColors.primaryColor
          }
        ];
      }
    };

    const getShadowStyle = () => {
      if (!customDisabled) {
        return [
          style.buttonWithShadow,
          { shadowColor: theme.themedColors.primaryBackground }
        ];
      }
    };

    return (
      <TouchableOpacity
        onPress={shouldShowProgressBar ? undefined : onPress}
        disabled={customDisabled}
        style={[
          style.button,
          getShadowStyle(),
          getButtonStyle(),
          {
            backgroundColor: theme.themedColors.primaryColor
          },
          customDisabled
            ? buttonDisableStyle ?? [
                buttonStyle,
                { backgroundColor: theme.themedColors.interface[400] }
              ]
            : buttonStyle
        ]}>
        <View testID="button-container" style={style.viewContainer}>
          {leftIcon && !shouldShowProgressBar && (
            <View
              style={[
                iconStyle,
                style.leftIconContainer,
                !shouldAlignTextWithLeftIconWithFullWidth
                  ? style.leftIconContainerPosition
                  : undefined
              ]}>
              {leftIcon?.(theme.themedColors.primaryColor, 20, 20)}
            </View>
          )}
          <View
            style={[
              style.textWithLoader,
              !shouldAlignTextWithLeftIconWithFullWidth
                ? style.textWithLoaderFlex
                : undefined,
              textContainerStyle
            ]}>
            {!shouldShowProgressBar && (
              <AppLabel
                style={[
                  style.text,
                  customDisabled ? textDisableStyle : textStyle
                ]}
                text={text}
                textType={textType}
                allowFontScaling={false}
              />
            )}
            {shouldShowProgressBar && (
              <ActivityIndicator
                testID="loader"
                style={[style.loader]}
                size={loaderSize}
                color={
                  loaderColor ?? theme.themedColors.secondaryBackground
                }
              />
            )}
          </View>
          {rightIcon && !shouldShowProgressBar && (
            <View style={style.rightIconContainer}>
              {rightIcon?.(theme.themedColors.primaryColor, 20, 20)}
            </View>
          )}
        </View>
      </TouchableOpacity>
    );
  }
);

const style = StyleSheet.create({
  buttonWithShadow: {
    elevation: 1,
    shadowOffset: { width: 0, height: 3 },
    shadowRadius: 3,
    shadowOpacity: 0.15
  },
  text: {
    fontSize: FONT_SIZE.base,
    overflow: "hidden",
    color: Colors.colors.white
  },
  loader: {
    marginLeft: 10
  },
  textWithLoader: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center"
  },
  textWithLoaderFlex: {
    flex: 1
  },
  leftIconContainer: {
    alignItems: "flex-start",
    marginLeft: SPACE.sm
  },
  leftIconContainerPosition: {
    position: "absolute",
    left: SPACE.sm
  },
  leftIcon: {
    marginLeft: 10,
    width: 20,
    height: 20
  },
  rightIconContainer: {
    position: "absolute",
    right: SPACE.lg,
    justifyContent: "flex-end",
    alignItems: "flex-end"
  },
  rightIcon: {
    marginRight: 10,
    width: 20,
    height: 20
  },
  buttonWithBorder: {
    borderWidth: 1,
    borderStyle: "solid"
  },
  buttonWithDash: {
    borderWidth: 1,
    borderStyle: "dashed"
  },
  button: {
    alignSelf: "center",
    flexDirection: "row",
    width: "100%",
    minHeight: 44,
    maxHeight: 55,
    justifyContent: "space-between",
    alignItems: "center",
    borderRadius: 25
  },
  viewContainer: {
    flexDirection: "row",
    alignItems: "center",
    padding: SPACE.sm
  }
});
