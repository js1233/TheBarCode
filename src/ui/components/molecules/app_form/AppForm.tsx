import { Formik, FormikValues } from "formik";
import React from "react";

export interface AppFormProps {
  children?: React.ReactNode;
  initialValues: FormikValues;
  onSubmit: (values: FormikValues) => void;
  validateOnMount?: boolean;
  validationSchema: FormikValues;
}

type Props = AppFormProps;

const AppForm: React.FC<Props> = ({
  initialValues,
  onSubmit,
  validationSchema,
  validateOnMount,
  children
}) => {
  /*  AppLog.logForcefullyForComplexMessages(
      () =>
    "AppForm => initialValues " + JSON.stringify(initialValues)
  );*/

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmit}
      validateOnMount={validateOnMount}
      validationSchema={validationSchema}>
      {() => <>{children}</>}
    </Formik>
  );
};

export default AppForm;
