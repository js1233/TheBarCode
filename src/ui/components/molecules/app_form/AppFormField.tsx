import { FONT_SIZE, SPACE } from "config/Dimens";
import { FormikValues, useFormikContext } from "formik";
import { usePreferredTheme } from "hooks";
import EIntBoolean from "models/enums/EIntBoolean";
import React, { useCallback, useEffect } from "react";
import { StyleSheet, View } from "react-native";
import { TouchableOpacity } from "react-native-gesture-handler";
import {
  AppLabel,
  AppLabelProps,
  TEXT_TYPE
} from "ui/components/atoms/app_label/AppLabel";
import {
  AppInputField,
  AppInputFieldProps
} from "../appinputfield/AppInputField";
import { AppFormValidationLabel } from "./AppFormValidationLabel";

export interface AppFormFieldProps {
  fieldTestID?: string;
  validationLabelTestID?: string;
  name: string;
  readOnly?: boolean;
  value?: string;
  labelProps?: AppLabelProps;
  fieldInputProps: AppInputFieldProps;
  linkLabelProps?: AppLabelProps;
  linkLabelOnPress?: () => void;
  secureTextEntry?: boolean;
  shouldNotOptimize?: boolean;
  customTextChanged?: (value: string) => void;
  isLocked?: EIntBoolean;
  isFieldInHorizontalContainer?: boolean;
}

type Props = AppFormFieldProps;

const AppFormField = React.memo<Props>(
  ({
    name,
    labelProps,
    fieldInputProps,
    fieldTestID,
    validationLabelTestID,
    linkLabelProps,
    linkLabelOnPress,
    secureTextEntry,
    customTextChanged,
    isLocked = EIntBoolean.FALSE,
    isFieldInHorizontalContainer = false
  }) => {
    const theme = usePreferredTheme();
    const { viewStyle, ...restFieldInputProps } = fieldInputProps;
    const {
      errors,
      setFieldTouched,
      setFieldValue,
      touched,
      initialValues
    } = useFormikContext<FormikValues>();

    const _setFieldTouched = useCallback(
      () => setFieldTouched(name),
      [setFieldTouched, name]
    );
    //I have used this for date of birth
    useEffect(() => {
      if (fieldInputProps.value) {
        setFieldValue(name, fieldInputProps.value);
        setTimeout(() => {
          _setFieldTouched();
        }, 200);
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [fieldInputProps.value]);

    const getView = () => {
      return (
        <>
          {labelProps && (
            <View style={styles.linkLabelContainer}>
              <AppLabel
                textType={TEXT_TYPE.BOLD}
                {...labelProps}
                style={[
                  styles.label,
                  { color: theme.themedColors.interface["900"] },
                  labelProps ? labelProps.style : {}
                ]}
              />
              <View style={styles.space} />

              {linkLabelProps && (
                <TouchableOpacity onPress={linkLabelOnPress}>
                  <AppLabel
                    textType={TEXT_TYPE.BOLD}
                    style={[
                      styles.linkLabel,
                      { color: theme.themedColors.primaryColor }
                    ]}
                    {...linkLabelProps}
                  />
                </TouchableOpacity>
              )}
            </View>
          )}
          <AppInputField
            testID={fieldTestID}
            valueToShowAtStart={initialValues[name]}
            onChangeText={(text) => {
              setFieldValue(name, text);
              _setFieldTouched();
              customTextChanged?.(text);
            }}
            onBlur={_setFieldTouched}
            secureTextEntry={secureTextEntry}
            viewStyle={[
              styles.inputFieldContainer,
              {
                backgroundColor: theme.themedColors.primaryBackground,
                borderColor: theme.themedColors.borderColor
              },
              viewStyle
            ]}
            {...restFieldInputProps}
            editable={!isLocked}
          />
          {errors[name] && touched[name] && (
            <AppFormValidationLabel
              validationLabelTestID={validationLabelTestID}
              errorString={errors[name] as string}
              shouldVisible={true}
            />
          )}
        </>
      );
    };

    return (
      <>
        {isFieldInHorizontalContainer ? (
          <View style={{ flexDirection: "column" }}>{getView()}</View>
        ) : (
          getView()
        )}
      </>
    );
  }
);

const styles = StyleSheet.create({
  label: {
    marginBottom: SPACE.xs,
    fontSize: FONT_SIZE.sm
  },
  space: {
    flex: 1
  },
  linkLabelContainer: {
    flexDirection: "row"
  },
  linkLabel: {
    fontSize: FONT_SIZE.xs
  },
  inputFieldContainer: { height: 44 }
});

export default AppFormField;
