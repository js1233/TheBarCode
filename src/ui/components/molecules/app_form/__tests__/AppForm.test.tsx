import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import AppForm from "ui/components/molecules/app_form/AppForm";
import * as Yup from "yup";
import AppFormField from "ui/components/molecules/app_form/AppFormField";
import { StyleSheet, View } from "react-native";
import { FormikValues } from "formik";
import AppFormFormSubmit from "ui/components/molecules/app_form/AppFormSubmit";
import { BUTTON_TYPES } from "ui/components/molecules/app_button/AppButton";

const validationSchema = Yup.object().shape({
  email: Yup.string()
    .required("Enter an email address")
    .email("Enter a valid email address"),
  password: Yup.string()
    .required("Enter your password")
    .min(8, "Password should be atleast 8 chars")
});

const TestingFormView = ({
  initialValues = { email: "", password: "" },
  onSubmit = (_values: FormikValues) => {}
}) => {
  return (
    <AppForm
      initialValues={initialValues}
      onSubmit={onSubmit}
      validationSchema={validationSchema}>
      <View style={styles.container}>
        <AppFormField
          fieldTestID="emailField"
          validationLabelTestID={"emailValidationLabel"}
          name="email"
          labelProps={{ text: "EMAIL" }}
          fieldInputProps={{
            textContentType: "emailAddress",
            keyboardType: "email-address",
            placeholder: "Enter your email address",
            autoCapitalize: "none"
          }}
        />
        <View style={styles.spacer} />
        <AppFormField
          fieldTestID="passwordField"
          validationLabelTestID="passwordValidationLabel"
          name="password"
          labelProps={{ text: "PASSWORD" }}
          fieldInputProps={{
            textContentType: "password",
            placeholder: "Enter your password",
            secureTextEntry: true
          }}
        />
        <View style={styles.spacer} />
        <AppFormFormSubmit
          testID={"formSubmit"}
          text={"Submit"}
          buttonType={BUTTON_TYPES.NORMAL}
        />
      </View>
    </AppForm>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 16.0
  },
  spacer: {
    padding: 8
  }
});

describe("rendering...", () => {
  const handleSubmit = jest.fn();
  const wrapper = render(
    <TestingFormView
      initialValues={{ email: "", password: "" }}
      onSubmit={handleSubmit}
    />
  );

  it("renders correctly", () => {
    const snapshot = wrapper.toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});

describe("validation testing", () => {
  const handleSubmit = jest.fn();
  it("should fail validation", async () => {
    const { getByTestId, getByText } = render(
      <TestingFormView
        initialValues={{ email: "", password: "" }}
        onSubmit={handleSubmit}
      />
    );

    const emailField = await getByTestId("emailField");
    const passwordField = await getByTestId("passwordField");

    const submitButton = await getByText("Submit");

    fireEvent.changeText(emailField, "test @.coa");
    fireEvent.changeText(passwordField, "");

    fireEvent.press(submitButton);

    await waitFor(() => {
      expect(getByTestId("emailValidationLabel")).toBeDefined();
      expect(getByTestId("passwordValidationLabel")).toBeDefined();
      expect(handleSubmit).not.toHaveBeenCalled();
    });
  });

  it("should pass validations", async () => {
    const { getByTestId, getByText } = render(
      <TestingFormView
        initialValues={{ email: "", password: "" }}
        onSubmit={handleSubmit}
      />
    );

    fireEvent.changeText(getByTestId("emailField"), "test@test.com");
    fireEvent.changeText(getByTestId("passwordField"), "12345678");

    fireEvent.press(getByText("Submit"));

    await waitFor(() => {
      expect(handleSubmit).toHaveBeenLastCalledWith(
        { email: "test@test.com", password: "12345678" },
        expect.anything()
      );
    });
  });
});
