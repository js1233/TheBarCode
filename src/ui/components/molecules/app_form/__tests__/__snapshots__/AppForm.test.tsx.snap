// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering... renders correctly 1`] = `
<View
  style={
    Object {
      "padding": 16,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontFamily": "ProximaNova-Bold",
          },
          Object {
            "color": "#171717",
            "fontSize": 27,
          },
          Object {
            "lineHeight": 38,
          },
          Array [
            Object {
              "fontSize": 27,
              "marginBottom": 6,
            },
            Object {
              "color": "#171717",
            },
            undefined,
          ],
        ]
      }
    >
      EMAIL
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "borderRadius": 6,
          "borderStyle": "solid",
          "color": "#000000",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        },
        Array [
          Object {
            "height": 44,
          },
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#D4D4D4",
          },
          undefined,
        ],
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      editable={true}
      keyboardType="email-address"
      multiline={false}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Enter your email address"
      placeholderTextColor="#134E4A"
      secureTextEntry={false}
      style={
        Array [
          Object {
            "color": "#000000",
            "flex": 1,
            "fontFamily": "ProximaNova-Regular",
            "fontSize": 25,
            "paddingHorizontal": 12,
            "textAlignVertical": "center",
          },
          undefined,
          Object {},
          null,
        ]
      }
      testID="emailField"
      textAlignVertical="center"
      textContentType="emailAddress"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "padding": 8,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontFamily": "ProximaNova-Bold",
          },
          Object {
            "color": "#171717",
            "fontSize": 27,
          },
          Object {
            "lineHeight": 38,
          },
          Array [
            Object {
              "fontSize": 27,
              "marginBottom": 6,
            },
            Object {
              "color": "#171717",
            },
            undefined,
          ],
        ]
      }
    >
      PASSWORD
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "borderRadius": 6,
          "borderStyle": "solid",
          "color": "#000000",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        },
        Array [
          Object {
            "height": 44,
          },
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#D4D4D4",
          },
          undefined,
        ],
      ]
    }
  >
    <TextInput
      editable={true}
      multiline={false}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Enter your password"
      placeholderTextColor="#134E4A"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "flex": 1,
            "fontFamily": "ProximaNova-Regular",
            "fontSize": 25,
            "paddingHorizontal": 12,
            "textAlignVertical": "center",
          },
          undefined,
          Object {},
          null,
        ]
      }
      testID="passwordField"
      textAlignVertical="center"
      textContentType="password"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "padding": 8,
      }
    }
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#0F766E",
        "borderRadius": 25,
        "elevation": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxHeight": 55,
        "minHeight": 44,
        "opacity": 1,
        "shadowColor": "#FFFFFF",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 3,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "padding": 8,
        }
      }
      testID="button-container"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "flex": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontFamily": "ProximaNova-Regular",
              },
              Object {
                "color": "#171717",
                "fontSize": 27,
              },
              Object {
                "lineHeight": 42,
              },
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 28,
                  "overflow": "hidden",
                },
                undefined,
              ],
            ]
          }
        >
          Submit
        </Text>
      </View>
    </View>
  </View>
</View>
`;
